sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -1597,59 +1590,71 @@ object SpillFramework extends Logging {
+ * `close` restores a bounce buffer to the pool for other callers to use.
  *
- * `release` actually closes the underlying DeviceMemoryBuffer, and should be called
+ * `release` actually closes the underlying buffer, and should be called
  * once at the end of the lifetime of the executor.
  *
- * @param dmb - actual cudf DeviceMemoryBuffer that this class is protecting.
+ * @param buf - actual cudf DeviceMemoryBuffer that this class is protecting.
+ * @param pool - the pool to which this buffer belongs
  */
-private[spill] case class DeviceBounceBuffer(var dmb: DeviceMemoryBuffer) extends AutoCloseable {
-  private var acquired: Boolean = false
-  def acquire(): DeviceBounceBuffer = synchronized {
-    while (acquired) {
-      wait()
-    }
-    acquired = true
-    this
-  }
-
-  private def unaquire(): Unit = synchronized {
-    acquired = false
-    notifyAll()
-  }
-
+private[spill] class BounceBuffer[T <: AutoCloseable](
+                                                       var buf: Option[T],
+                                                       private val pool: BounceBufferPool[T])
+  extends AutoCloseable {
   override def close(): Unit = {
-    unaquire()
+    pool.returnBuffer(this)
   }
 
-  def release(): Unit = synchronized {
-    if (acquired) {
-      throw new IllegalStateException(
-        "closing device buffer pool, but some bounce buffers are in use.")
-    }
-    if (dmb != null) {
-      dmb.close()
-      dmb = null
-    }
+  def release(): Unit = {
+    buf.foreach(_.close())
+    buf = None
   }
 }
 
+
 /**
- * A bounce buffer pool with buffers of size `bufferSize`
+ * A bounce buffer pool with buffers of size `bufSize`
  *
- * This pool returns instances of `DeviceBounceBuffer`, that should
+ * This pool returns instances of `BounceBuffer[T]`, that should
  * be closed in order to be reused.
  *
  * Callers should synchronize before calling close on their `DeviceMemoryBuffer`s.
  */
-trait DeviceBounceBufferPool extends AutoCloseable {
-  def bufferSize: Long
-  def nextBuffer(): DeviceBounceBuffer
+class BounceBufferPool[T <: AutoCloseable](private val bufSize: Long,
+                                           private val bbCount: Long,
+                                           private val allocator: Long => T) extends AutoCloseable {
+  private val pool = new ConcurrentLinkedQueue[BounceBuffer[T]]
+  for (_ <- 1L to bbCount) {
+    pool.offer(new BounceBuffer[T](Some(allocator(bufSize)), this))
+  }
+
+  def bufferSize: Long = bufSize
+  def nextBuffer(): BounceBuffer[T] = synchronized {
+    var res = pool.poll()
+    while (res == null) {
+      wait()
+      res = pool.poll()
+    }
+
+    res
+  }
+  def returnBuffer(buffer: BounceBuffer[T]): Unit = synchronized {
+    pool.offer(buffer)
+    notifyAll()
+  }
+
+  override def close(): Unit = {
+    if (pool.size() < bbCount) {
+      throw new IllegalStateException("tried to close BounceBufferPool when buffers are still " +
```

Comment:
Just for being extra robust, could we log an error, set a flag that this is now closed, and then when anyone returns a buffer could we release it if we are closed instead of offering it back to the pool?